### ISSUE TITLE
[no-implied-eval] False positive with callback: Function

### ISSUE THREAD

**Issue by tosmolka** [typescript-eslint/typescript-eslint#2358]
**Repro**

```JavaScript
module.exports = {
    "env": {
        "browser": true,
        "es2020": true
    },
    "parser": "@typescript-eslint/parser",
    "parserOptions": {
        "ecmaVersion": 11,
        "sourceType": "module",
        "project": "tsconfig.json"
    },
    "plugins": [
        "@typescript-eslint"
    ],
    "rules": {
        "@typescript-eslint/no-implied-eval": "error"
    }
};
```

```TS
const foo = (callback: Function) => {
    setTimeout(callback, 0);
};
```

**Expected Result**
ESLint detects 0 problems

**Actual Result**
  2:16  error  Implied eval. Consider passing a function  @typescript-eslint/no-implied-eval

✖ 1 problem (1 error, 0 warnings)

**Additional Info**

Handler argument is parsed as `Identifier` but `isFunctionType()` returns for it false. 

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.8.0` |
| `@typescript-eslint/parser`        | `3.8.0` |
| `TypeScript`                       | `3.9.7` |
| `ESLint`                           | `7.6.0` |
| `node`                             | `14.6.0` |
| `npm`                              | `6.14.5` |

**Comment by bradzacher** [comment]
Why are you using the `Function` type here? It's generally recommended that you _don't_ use this type ever, as it has a distinct lack of clarity as to what it means.

The bug here is that the `Function` symbol is not "function-like", and the `Function` type has no call-signatures.

If you use something like `(...args: any[]) => void` instead, then it will work fine.

**Comment by tosmolka** [comment]
Thanks @bradzacher , I don't own the code that is causing those false positives so can't comment on that. My assumption was that the rule should identify calls to `setTimeout` with string argument and this seemed like a bug.

**Comment by bradzacher** [comment]
This is a bug! The `Function` type is unique in that in terms of the underlying type data it both is and isn't a function at the same time.

We have to add a special case for it.

Note that this rule doesn't ban string arguments, it instead bans non-functional arguments.

The reason it works that way is because under the hood, functions like `setTimeout` actually coerce any non-function input to string.

For example

```
setTimeout(["console.log('hihihi')"], 5)
````

Will coerce the array of one element to a string which will just be the string itself and log a line.
